### ISSUE TITLE
benign-looking 'replace' causes Qed to blow up

### ISSUE THREAD

**Issue by andrew-appel** [coq/coq#12378]
#### Description of the problem

In Coq 8.11.1, a 'replace' command causes the Qed to blow up (I killed it after 1 hour), but without the 'replace', the Qed takes about 0.2 seconds.

That is:  In my theorem, after 100 lines of nontrivial proof script (but the kind that is normally well within Coq's capacity), I come to this proof goal:

```
j, c : forall _ : environ, mpred
H7 : derives c j
______________________________________(1/1)
True
```

If I finish the proof with  `auto. Qed.`   then the Qed takes 0.22 seconds.
If I finish the proof with `replace c with j in H7 by apply admit. auto. Qed.`  then the Qed takes more than an hour, if indeed it ever finishes.  

The fact that I'm using a very generous Admitted lemma called `admit` is irrelevant, I believe.  The original version from which I distilled this did not play any such trick.

You can reproduce this as follows:  
1. Download commit [f2549d0](https://github.com/PrincetonUniversity/VST/commit/f2549d0d66ef99911e12acbd0c52abedaa5ea92f) of the VST repository, 
2. Do "make -j7"  (or whatever -j argument you like)
3.  In the last-built file, progs/verif_binary_heap.v, find the line near the end labeled 
 `(* UNCOMMENT THIS LINE TO HAVE Qed BLOW UP IN Coq 8.11.1 ... *)`
and uncomment the `replace` command.
4.  Notice that Qed now blows up.  Memory usage doesn't blow up, just run time.

**Comment by JasonGross** [comment]
I see that shortly above your line you have `simpl eval_binop in H7` at https://github.com/PrincetonUniversity/VST/blob/f2549d0d66ef99911e12acbd0c52abedaa5ea92f/progs/verif_binary_heap.v#L584  Furthermore, none of the lines between that and `replace` introduce cast nodes in the proof (basically you are just posing variables, which does not impact typing, and running `change` (via things like `set`), which leaves no evidence in the proof term.  However, the `replace` line leaves evidence in the proof term, and forces `Qed` to check `H7` against its new simplified type.  I strongly suspect that this issue is just "`simpl` doesn't leave any evidence in the proof term, causing `Qed` to blow up because it uses a different conversion strategy".

**Comment by andrew-appel** [comment]
It's definitely not the `simpl eval_binop in H7` that causes the problem, but it could be some earlier `simpl`.

**Comment by JasonGross** [comment]
@andrew-appel How do you know that the problem shows up before line 522?

Btw, if you do
```diff
 simpl eval_binop in H7.
-match type of H7 with _ |-- ?A => set (j:=A) in H7 end.
-clearbody j.
-set (c := andp _ _) in *.
-clearbody c.
-clear - H7.
-(* UNCOMMENT THIS LINE TO HAVE Qed BLOW UP IN Coq 8.11.1 
-    replace c with j in H7 by apply admit.  *)
+pose proof H7.
```
is that enough to cause the slowdown?

**Comment by andrew-appel** [comment]
Yes, with `pose proof H7` instead of those other lines, it still blows up.

The reason I know the problem is already lurking before line 522 is this:  The original proof script that had the Qed problem went to line 522, then I had a `forward_if ...` tactic and then other stuff.  In diagnosing what caused the problem, I unfolded the `forward_if` into its component parts, recursively, until I found the primitive tactic `replace` that seemed to be causing the problems.  So therefore:  Probably, something before line 522 is causing a `replace` that happens as part of the scripting at line 522 to make the Qed blow up.

**Comment by JasonGross** [comment]
Does `forward_if` invoke `simpl`?  If so, I'd be 80% confident that the issue is `simpl` (`replace` just turns it into a problem that `Qed` can see rather than one it can't, much like `pose proof`).  As for why `forward_if` would make this proof blow up and not others, the reason lies in the details of the type of `H7` before you simplify it.

If you issue `Opaque derives.` or `Strategy 100 [derives].` (really you can use any positive number in `Strategy`, as long as you haven't already issued other strategy commands), does that fix the blowup at `Qed`?

**Comment by andrew-appel** [comment]
Yes indeed, the `simpl in v` at line 540 is part of `forward if`, in effect, and when I remove that line, the Qed goes fast.

Putting `Opaque derives` just before Qed does not help, and putting it much earlier can't work for me because some of my tactic scripting seems to rely on the transparency of derives.

`Strategy 100 [derives]` does not make a difference, even if I put it just before the `Lemma` command; the Qed still blows up.

**Comment by JasonGross** [comment]
> Yes indeed, the `simpl in v` at line 540 is part of `forward if`, in effect, and when I remove that line, the Qed goes fast.

So it sounds like we found the root of the issue.

> Putting `Opaque derives` just before Qed does not help, and putting it much earlier can't work for me because some of my tactic scripting seems to rely on the transparency of derives.
> 
> `Strategy 100 [derives]` does not make a difference, even if I put it just before the `Lemma` command; the Qed still blows up.

In that case, this is beyond my ability to debug further without having it running in Coq in front of me.  If you know exactly which constants should be unfolded by `simpl`, you can issue `Strategy -100 [list of constants always unfolded by simpl]`, and perhaps you'll also need to issue `Strategy -50 [list of constants only sometimes unfolded by simpl]` to deal with the fact that `simpl` might unfold some constants some of the time and not others (and you want to tell `Qed`: first unfold the constants always unfolded by `simpl`, then unfold the constants sometimes unfolded by `simpl`, and only after that unfold the constants never unfolded by `simpl`), and that should make `Qed` take approximately the same conversion path as `simpl`, give or take.

**Comment by andrew-appel** [comment]
I have adjusted VST's forward_if tactic so that `simpl v` is used in fewer subgoals, such that it doesn't seem to cause the Qed blowup.  Still, the underlying Coq problem is probably, indeed,  "simpl doesn't leave any evidence in the proof term, causing Qed to blow up because it uses a different conversion strategy".

**Comment by ejgallego** [comment]
@corwin-of-amber and I found an interesting case of this issue, where `simpl` on a term that is not simplified, `simpl` will insert a cast that I think it is useless. Example:

```coq
Lemma foo : True.
Proof.
Show Proof.
simpl.
Show Proof.
simpl.
Show Proof.
Qed.
```
The first `Show Proof` will print `?Goal` , the last one will print `((?Goal : True) : True)` , so two casts are added.

There is an additional problem here, is that the kernel will easily indeed stack-overflow when checking a nested sequence of casts. You can try with `Ltac simpl_of_hanoi := do N simpl.` in the above script.

**Comment by ppedrot** [comment]
> that the kernel will easily indeed stack-overflow when checking a nested sequence of casts

I can't reproduce that. On my machine, your example with 100,000 useless casts doesn't trigger any stack overflow. I believe that by that level of nesting, you deserve any kind of stack overflow, since the same kind of issue happens with many similar nested terms (there is a reason why we warn against not-so-big unary natural numbers). @ejgallego  Aren't you by any chance using the js target which is more prone to raising overflows?

**Comment by ejgallego** [comment]
I am using Coq native, I needed to insert around 90k casts to make that happen, but indeed, that's pathological case.
